### PR TITLE
First draft of the TreeHandler

### PR DESCRIPTION
**Pull request by fmazzasc**
Hi guys, following the work done by @mpuccio in his fork I have implemented a first draft of the `TreeHandler`. My idea is that this module should cover the steps that go from the `TTree` conversion to the `ModelHandler.train_test_model model` method. I tested It with small trees and seems to me that everything works fine both for the binary and the multiclass cases. Going into details this P.R provides:

- The class `TreeHandler` that converts a `TTree` into a `pandas.DataFrame`. The class includes getters, methods to apply preselections and to slice the DataFrame in different bins
- The method `train_test_split` implemented in `analysis_utils.py`. This is a wrapper to the sklearn method and takes as input a list containing `TreeHandlers` belonging to different classes( e.g. `[SignalTreeHandler, BackgroundTreeHandler]`)

Finally, since this is a quite important change, I think It would be important if everybody have a look and comment/approve/disapprove this P.R.
Thanks!

### CODE REVIEW

File: hipe4ml/analysis_utils.py [hunk 1]
```diff
@@ -110,9 +112,76 @@ class should be labeled with 0, ..., N.
...
+    """
+    Return a list containing respectively training set dataframe,
+    training label array, test set dataframe, test label array
+    computed from a list of TreeHandler objects. If sliced_df == True
+    the method preforms the train-test split for each slice
+
+    Input
+    -----------------------------------------------
+    data_list: list
+        List of TreeHandler models. For example: if ypu perform binary
```
Comment (lbariogl) [inline hipe4ml/analysis_utils.py]: ypu --> you

File: hipe4ml/analysis_utils.py [hunk 2]
```diff
@@ -110,9 +112,76 @@ class should be labeled with 0, ..., N.
...
+        classification the list should contain the TreeHandlers corresponding
```
Comment (lbariogl) [inline hipe4ml/analysis_utils.py]: classification,

File: hipe4ml/analysis_utils.py [hunk 3]
```diff
@@ -110,9 +112,76 @@ class should be labeled with 0, ..., N.
...
+        to the signal and the background candidates
+
+    labels_list: list
+        List containing the labels associated to each DataHandler. For example:
+        if you perform binary classification the list should be [1,0]
```
Comment (lbariogl) [inline hipe4ml/analysis_utils.py]: binary classification,

File: hipe4ml/analysis_utils.py [hunk 4]
```diff
@@ -110,9 +112,76 @@ class should be labeled with 0, ..., N.
...
+
+    sliced_df: bool
+        If True the function searches for the slices stored in the DataHandler
```
Comment (lbariogl) [inline hipe4ml/analysis_utils.py]: If True,

File: hipe4ml/tree_handler.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        Apply preselections to data
+
+        Input
+        ------------------------------------------------
+        preselection: str
+            String containing the cuts to be applied as preselection on the data contained in the original
+            tree. The string syntax is the one required in the pandas.DataFrame.query() method.
+            You can refer to variables in the environment by prefixing them with an ‘@’ character like @a + b.
+        inplace: bool
+            If True the preselected dataframe replaces the initial dataframe. Otherwise return a copy of the
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: If True``,`` ... Otherwise return``s``...

File: hipe4ml/tree_handler.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+            self._full_data_frame = self._full_data_frame.query(
+                self._preselections)
+            return df_pres
+
+        df_pres = self._full_data_frame.query(preselections)
+        return df_pres
+
+    def slice_data_frame(self, projection_variable, projection_binning, delete_original_df=False):
+        """
+        Create a list containing slices of the orginal DataFrame.
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: Create``s``

File: hipe4ml/tree_handler.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+        Input
+        ------------------------------------------------
+        projection_variable: str
+            Name of the variable that will be sliced in the analysis
+        projection_binning: list
+            Binning of the sliced variable should be given as a list of
+            [min, max) values for each bin
+
+        delete_original_df: bool
+            If True delete the original DataFrame. Only the
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: If true delete``s``

File: hipe4ml/tree_handler.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+            bin_mask = np.logical_and(
+                self._full_data_frame[projection_variable] >= ibin[0],
+                self._full_data_frame[projection_variable] < ibin[1])
+            self._sliced_df_list.append(self._full_data_frame[bin_mask])
+        if delete_original_df:
+            self._full_data_frame = None
+
+    def shuffle_data_frame(self, size=None, frac=None, inplace=True):
+        """
+        Compute a random sample of the DataFrame
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: ``Extracts`` a random sample ```from```

File: hipe4ml/tree_handler.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+        size: int
+            Number of candidates to return. Cannot be used with frac. Default = 1 if
+            frac = None.
+
+        frac: float
+            Fraction of candidates to return. Cannot be used with n.
+
+        inplace: bool
+            If True the shuffled dataframe replaces the initial dataframe. Otherwise return a copy
+            of the shuffled df
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: Otherwise return``s`` ...

File: hipe4ml/tree_handler.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+        Output
+        df: pandas.DataFrame or None
+            if inplace == True returns the shuffled DataFrame
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: As above, if inplace==True it returns None and replaces the full data frame

File: hipe4ml/tree_handler.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        if inplace:
+            self._full_data_frame = self._full_data_frame.sample(size, frac)
+            return df_shuf
+
+        df_shuf = self._full_data_frame.sample(size, frac)
+        return df_shuf
+
+    def eval_data_frame(self, ev_str, inplace=True):
+        """
+        Evaluate a string describing operations on DataFrame columns
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: Evaluate``s`` ...

File: hipe4ml/tree_handler.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+
+        Input
+        ------------------------------------------------
+        preselection: str
+            The expression string to evaluatedatetime A combination of a date and a time. Attributes: ()
+            The string syntax is the one required in the pandas.DataFrame.eval() method.
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: is ``that`` ...

File: hipe4ml/tree_handler.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        inplace: bool
+            If the expression contains an assignment, whether to perform the operation inplace and mutate
+            the existing DataFrame. Otherwise, a new DataFrame is returned.
+
+        Output
+        df: pandas.DataFrame or None
+            if inplace == True returns the new DataFrame
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: if inplace == True, None is returned and the full dataframe is evaluated

File: hipe4ml/tree_handler.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+        if inplace:
+            self._full_data_frame.eval(ev_str, inplace=True)
+            return df_ev
+
+        df_ev = self._full_data_frame.eval(ev_str)
+        return df_ev
+
+    def print_infos(self):
+        """
+        Print informations about the DataHandler object and its
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: Print``s the``  information~~s~~ (uncountable)

File: hipe4ml/tree_handler.py [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+        print("\nFile name: ", self._file)
+        print("Tree name: ", self._tree)
+        print("DataFrame head:\n", self._full_data_frame.head(5))
+        print("\nPreselections:", self._preselections)
+        print("Sliced variable: ", self._projection_variable)
+        print("Slices binning: ", self._projection_binning)
+
+    def write_df_to_parquet_files(self, base_file_name="TreeDataFrame", path="./", save_slices=False):
+        """
+        Write the pandas dataframe to parquet files
```
Comment (lbariogl) [inline hipe4ml/tree_handler.py]: Write``s``

**Comment by fmazzasc** [comment]
Hi @lbariogl, I have solved all your comments. Thanks!

File: hipe4ml/tree_handler.py [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+        Parameters
+        ------------------------------------------------
+        file_name: str
+            Name of the input file where the data sit
+        tree_name: str
+            Name of the tree within the input file
+        columns_name: list
+            List of the names of the branches that one wants to analyse
+        """
+        self._file = uproot.open(file_name)
```
Comment (fgrosa) [inline hipe4ml/tree_handler.py]: It would be nice to have the possibility to load parquet files besides root files (for example @fcatalan92 and I typically work with parquet files in input). You could for example check the extension of the input file and use `uproot` in case of `ROOT` files and `pandas.read_parquet` in case of `parquet` files

File: hipe4ml/tree_handler.py [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+        self._projection_binning = projection_binning
+
+        self._sliced_df_list = []
+        for ibin in projection_binning:
+            bin_mask = np.logical_and(
+                self._full_data_frame[projection_variable] >= ibin[0],
+                self._full_data_frame[projection_variable] < ibin[1])
+            self._sliced_df_list.append(self._full_data_frame[bin_mask])
+        if delete_original_df:
+            self._full_data_frame = None
```
Comment (fgrosa) [inline hipe4ml/tree_handler.py]: I am not sure whether this really deletes the dataframe (if it is the case, just ignore this comment). Wouldn't be better to use 
```python
del self._full_data_frame
``` 
instead?

**Comment by fmazzasc** [comment]
Hi @fgrosa thanks a lot for the comments. I have added the possibility to initialize the `TreeHandler` starting from a .parquet file

File: hipe4ml/tree_handler.py (hunk 17 above)
Comment (fmazzasc) [reply hipe4ml/tree_handler.py]: I cross-checked that and I'm almost sure that assigning self._full_data_frame = None is enough to delete it!

File: hipe4ml/tree_handler.py [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+import uproot
+
+
+class TreeHandler:
+    """
+    Class for storing and managing the data of a ROOT tree from a .root file
+    or a pandas.DataFrame from a .parquet file
+    """
+
+    def __init__(self, file_name, tree_name=None, columns_names=None, **kwds):
```
Comment (mpuccio) [inline hipe4ml/tree_handler.py]: For both parquet and ROOT having the default columns_names =None might not be optimal. I would rather have the wildcard that reads all the columns in both cases.

Comment (fmazzasc) [reply hipe4ml/tree_handler.py]: Hi @mpuccio, the method already implements that but i forgot to specify It. Thanks!

File: hipe4ml/tree_handler.py [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+            String containing the cuts to be applied as preselection on the data contained in the original
+            tree. The string syntax is the one required in the pandas.DataFrame.query() method.
+            You can refer to variables in the environment by prefixing them with an ‘@’ character like @a + b.
+        inplace: bool
+            If True, the preselected dataframe replaces the initial dataframe. Otherwise return a copy of the
+            preselected df
+
+        Returns
+        ------------------------------------------------
+        out: pandas.DataFrame or None
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: I think it is better to keep the names used in the code, this is what scikit-learn does for example.

File: hipe4ml/tree_handler.py [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+        frac: float
+            Fraction of candidates to return. Cannot be used with n.
+
+        inplace: bool
+            If True the shuffled dataframe replaces the initial dataframe. Otherwise return a copy
+            of the shuffled df
+
+        Returns
+        ------------------------------------------------
+        out: pandas.DataFrame or None
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: See comment at line 142.

File: hipe4ml/tree_handler.py [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+            If inplace == True None is returned and the full DataFrame is replaced
+
+        """
+        df_shuf = None
+        if inplace:
+            self._full_data_frame = self._full_data_frame.sample(size, frac)
+            return df_shuf
+
+        df_shuf = self._full_data_frame.sample(size, frac)
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: It could be useful to be able to change the seed of the sampling, maybe some variables/kwargs of sample can be exposed in the function.

File: hipe4ml/tree_handler.py [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+        return df_shuf
+
+    def eval_data_frame(self, ev_str, inplace=True):
+        """
+        Evaluate a string describing operations on DataFrame columns
+
+        Parameters
+        ------------------------------------------------
+        preselection: str
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: preselection -> ev_str

File: hipe4ml/tree_handler.py [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+            The expression string to evaluate. A combination of a date and a time. Attributes: ()
+            The string syntax is the one required in the pandas.DataFrame.eval() method.
+        inplace: bool
+            If the expression contains an assignment, whether to perform the operation inplace and
+            mutate the existing DataFrame. Otherwise, a new DataFrame is returned.
+
+        Returns
+        ------------------------------------------------
+        out: pandas.DataFrame or None
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: See comment at line 142.

File: hipe4ml/tree_handler.py [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+        return df_shuf
+
+    def eval_data_frame(self, ev_str, inplace=True):
+        """
+        Evaluate a string describing operations on DataFrame columns
+
+        Parameters
+        ------------------------------------------------
+        preselection: str
+            The expression string to evaluate. A combination of a date and a time. Attributes: ()
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: Do `A combination of a date and a time. Attributes: ()` means something?

File: hipe4ml/tree_handler.py [hunk 25]
```diff
@@ -0,0 +1,278 @@
...
+        if inplace:
+            self._full_data_frame.eval(ev_str, inplace=True)
+            return df_ev
+
+        df_ev = self._full_data_frame.eval(ev_str)
+        return df_ev
+
+    def print_summary(self):
+        """
+        Print information about the DataHandler object and its
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: DataHandler -> TreeHandler

File: hipe4ml/tree_handler.py [hunk 26]
```diff
@@ -0,0 +1,278 @@
...
+        ------------------------------------------------
+        base_file_name: str
+            Base filename used to save the parquet files
+        path: str
+            Base path of the output files
+        save_slices: bool
+            If True and the slices are available, single parquet files for each
+            bins are created
+        """
+        self._full_data_frame.to_parquet(
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: What happens here if the original dataframe has been deleted by `slice_data_frame(..)` ?

File: hipe4ml/tree_handler.py [hunk 27]
```diff
@@ -0,0 +1,278 @@
...
+            f"{path}{base_file_name}.parquet.gzip", compression="gzip")
+        if save_slices:
+            for ind, i_bin in enumerate(self._projection_binning):
+                name = f"{path}{base_file_name}_{self._projection_variable}_{i_bin[0]}_{i_bin[1]}"
+                self._sliced_df_list[ind].to_parquet(f"{name}.parquet.gzip",
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: And here if there are no slices in the object?

**Comment by fmazzasc** [comment]
Hi @fcatalan92, thanks for the comments! I solved most of them and I also removed the blank lines before the end of the docstring

File: hipe4ml/tree_handler.py [hunk 28]
```diff
@@ -136,22 +134,15 @@ def apply_preselections(self, preselections, inplace=True):
...
-        df_pres = None
-        if inplace:
-            self._preselections = preselections
-            self._full_data_frame = self._full_data_frame.query(
-                self._preselections)
-            return df_pres
-
-        df_pres = self._full_data_frame.query(preselections)
-        return df_pres
+        return self._full_data_frame.query(preselections, inplace=inplace, **kwds)
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: Are you sure that this works also for ```inplace=True```? It looks to me that the pandas query function does not return a value when performing the operation in place https://github.com/pandas-dev/pandas/blob/v1.0.3/pandas/core/frame.py#L3083-L3243.

In addition, I think that storing the selection string in ```self._preselection``` when performing the operation in place could be useful.

File: hipe4ml/tree_handler.py [hunk 29]
```diff
@@ -201,51 +191,47 @@ def shuffle_data_frame(self, size=None, frac=None, inplace=True):
...
-
         """
-        df_ev = None
-        if inplace:
-            self._full_data_frame.eval(ev_str, inplace=True)
-            return df_ev
-
-        df_ev = self._full_data_frame.eval(ev_str)
-        return df_ev
+        return self._full_data_frame.eval(ev_str, inplace=inplace, **kwds)
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: Same consideration here as for the query function.

File: hipe4ml/tree_handler.py [hunk 30]
```diff
@@ -269,10 +255,14 @@ def write_df_to_parquet_files(self, base_file_name="TreeDataFrame", path="./", s
             If True and the slices are available, single parquet files for each
             bins are created
         """
-        self._full_data_frame.to_parquet(
-            f"{path}{base_file_name}.parquet.gzip", compression="gzip")
+        if self._full_data_frame is not None:
+            self._full_data_frame.to_parquet(
+                f"{path}{base_file_name}.parquet.gzip", compression="gzip")
```
Comment (fcatalan92) [inline hipe4ml/tree_handler.py]: I think we can add a warning when there is no dataframe to save, like what is done for the slices.

**Comment by fcatalan92** [comment]
Hi @fmazzasc, it looks good. I have only some minor questions/suggestions.
